python/lsst/ts/tunablelaser/csc.py

```diff
@@ -0,0 +1,261 @@
+            self.model.run()
+        except Exception as e:
+            self.evt_errorCode(
+                errorCode=TunableLaser.LaserErrorCode.general_error,
+                errorReport=e.msg,
+                traceback=traceback.format_exc()
+            )
+        self.detailed_state = TunableLaser.LaserDetailedState.PROPAGATINGSTATE
+
+    async def do_stopPropagateLaser(self, data):
+        """Stop the Propagating State of the laser.
+
+        Parameters
+        ----------
+        data
+        """
+        self.assert_enabled("stopPropagateLaser")
+        self.assert_substate([TunableLaser.LaserDetailedState.PROPAGATING], "stopPropagateLaser")
+        try:
+            self.model.stop()
+        except Exception as e:
+            self.evt_errorCode(
+                errorCode=TunableLaser.LaserErrorCode.general_error,
+                errorReport=e.msg,
+                traceback=traceback.format_exc()
+            )
+        self.detailed_state = TunableLaser.LaserDetailedState.ENABLEDSTATE
+
+    async def do_clearFaultState(self, data):
+        """Clear the fault state of the laser by turning the power register
+        off.
+
+        Parameters
+        ----------
+        data
+        """
+        self.model.clear_fault()
+
+    @property
+    def detailed_state(self):
+        """Return the current substate of the laser and when it changes
+        publishes an event.
+        """
+        return TunableLaser.LaserDetailedState(self.evt_detailedState.data.detailedState)
+
+    @detailed_state.setter
+    def detailed_state(self, new_sub_state):
+        new_sub_state = TunableLaser.LaserDetailedState(new_sub_state)
+        self.evt_detailedState.set_put(detailedState=new_sub_state)
+
+    async def begin_enable(self, data):
+        """A hook that sets up the laser for propagation so that it
+        is ready to go.
+
+        Parameters
+        ----------
+        data
+        """
+        try:
+            self.model._laser.MaxiOPG.set_configuration(self.model.csc_configuration.optical_configuration)
+            self.model._laser.set_output_energy_level("MAX")
+        except TimeoutError as te:
+            self.fault(code=TunableLaser.LaserErrorCode.timeout_error, report=te.msg)
+
+    async def begin_disable(self, data):
+        """
+
+        Parameters
+        ----------
+        data
+        """
+        if self.model._laser.M_CPU800.power_register_2.register_value == "ON":
+            try:
+                self.model.stop()
+                self.detailed_state = TunableLaser.LaserDetailedState(
+                    TunableLaser.LaserDetailedState.ENABLEDSTATE)
+            except TimeoutError as te:
+                self.fault(code=TunableLaser.LaserErrorCode.timeout_error, report=te.msg)
+        self.model.set_output_energy_level("OFF")
+
+    async def end_start(self, data):
+        try:
+            self.model.connect()
+            self.telemetry_task = asyncio.ensure_future(self.telemetry())
```

Comment:
You might consider moving this to handle_summary_state. Advantages:
* You don't have to worry about how the CSC gets into the state. Some states have two different commands to enter that state.
* It keeps related code together.

Combined with the strong recommendation for end_standby you end up with:
```
def handle_summary_state(self):
    if self.disabled_or_enabled:
        # I prefer to build the model and/or TCP/IP connections here.
        # This is especially helpful if the model and/or connections
        # depend on the configuration. To do that, use:
        # if self.model is None:
        #     self.model = ....
        # But your idiom is fine if the model does not depend on the config:
        if not self.model.connected():
             self.model.connect()
        if self.telemetry_task.done():
            self.telemetry_task = asyncio.create_task(self.telemetry())
    else:
        self.model.disconnect()
        # If you use my suggested idiom then also set the model to None:
        # self.model = None
        self.telemetry_task.cancel()
```

By the way: I plan to deprecate the begin_ and end_ methods in salobj soon, and eliminate them in salobj 6.